### PR TITLE
[GLUTEN-5771][VL] Add metrics for ColumnarArrowEvalPythonExec

### PR DESCRIPTION
**Pull request by yma11**
## What changes were proposed in this pull request?

Add metric for ColumnarArrowEvalPythonExec

(Fixes: \#5771)

## Spark UI

image

## How was this patch tested?

We tested performance of arrow udf and collected some performance:
```
from pyspark.sql.functions import pandas_udf, PandasUDFType
import pyspark.sql.functions as F
import os
@pandas_udf('long')
def pandas_plus_one(v):
    return (v + 1)

@pandas_udf('string')
def pd_get_first(v):
    return v.str.split(':').str[1]
# test int
df = spark.read.orc("file:///xx/yy").select("user_id").withColumn("processed_user_id", pandas_plus_one("user_id")).select("processed_user_id")
# test string
df = spark.read.orc("file:///xx/yy").select("url").withColumn("processed_url", pd_get_first("url")).select("processed_url")
```

The perf shows ~20% perf gain compared with vanilla spark.

image

### CODE REVIEW

**Comment by FelixYBW** [comment]
@yma11 can you add a UI chart for the pyarrow UDF? Also add some implementation details?

In theory we can convert Velox to Arrow in Velox pipeline, then pass the arrow pointer to Spark where it's send to python process. There is no C2R and R2C in the whole process and no memcpy between Velox and Spark. Can we achieve this?

**Comment by yma11** [comment]
> @yma11 can you add a UI chart for the pyarrow UDF? Also add some implementation details?
> 
> In theory we can convert Velox to Arrow in Velox pipeline, then pass the arrow pointer to Spark where it's send to python process. There is no C2R and R2C in the whole process and no memcpy between Velox and Spark. Can we achieve this?

Yes. There is no C2R and R2C in current implementation. There is a VeloxColumnar to Arrow only. But for memcpy, it depends on the arrow bridge. I found there are still some memory allocation at velox for data types like string. Let me add the implementation under the feature track.

**Comment by yma11** [comment]
> > @yma11 can you add a UI chart for the pyarrow UDF? Also add some implementation details?
> > In theory we can convert Velox to Arrow in Velox pipeline, then pass the arrow pointer to Spark where it's send to python process. There is no C2R and R2C in the whole process and no memcpy between Velox and Spark. Can we achieve this?
> 
> Yes. There is no C2R and R2C in current implementation. There is a VeloxColumnar to Arrow only. But for memcpy, it depends on the arrow bridge. I found there are still some memory allocation at velox for data types like string. Let me add the implementation under the feature track.

@FelixYBW The implementation details are now added in [5461](https://github.com/apache/incubator-gluten/issues/5461). Perf data is also wrapped there. FYI.

**Review by zhztheplayer — COMMENTED** [review]
I just noticed that this file's package is `package org.apache.spark.api.python` which is wrong. Would you like to fix it? @yma11

File: backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala [hunk 1]
```diff
@@ -341,11 +355,20 @@ case class ColumnarArrowEvalPythonExec(
               val joinedVectors = (0 until inputCb.numCols).toArray.map(
                 i => inputCb.column(i)) ++ (0 until outputCb.numCols).toArray.map(
                 i => outputCb.column(i))
+              // Columns in outputCb has random 0 or 1 refCnt and will fail checks in ensureOffload,
+              // so we do a hard reset here.
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala:359]: Did you figure out why "random ref count" happens?

File: gluten-data/src/main/java/org/apache/gluten/vectorized/ArrowWritableColumnVector.java [hunk 2]
```diff
@@ -345,6 +345,10 @@ public void retain() {
     refCnt.getAndIncrement();
   }
 
+  public void resetRefCnt(long value) {
+    refCnt.set(value);
+  }
```
Comment (zhztheplayer) [inline gluten-data/src/main/java/org/apache/gluten/vectorized/ArrowWritableColumnVector.java:350]: There was already a `#forceClose` API in `ColumnarBatches`, I assume that one could cover your use case.

Although let's figure out why we need to forcibly close the vectors first.

File: backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala (hunk 1 above)
Comment (yma11) [reply backends-velox/src/main/scala/org/apache/gluten/execution/python/ColumnarArrowEvalPythonExec.scala:359]: Not quite. The outputCb is directly retrieved from python worker so can't debug why it happens.

**Comment by yma11** [comment]
> I just noticed that this file (ColumnarArrowEvalPythonExec.scala)'s package is `package org.apache.spark.api.python` which is wrong. Would you like to fix it? @yma11

Fixed.

File: docs/developers/VeloxUDF.md [hunk 3]
```diff
@@ -179,3 +179,48 @@ The output from spark-shell will be like
 +------------------+----------------+
 ```
 
+# Pandas UDFs (a.k.a. Vectorized UDFs)
```
Comment (yma11) [inline docs/developers/VeloxUDF.md:182]: @zhouyuan Can you take a look at this contents for pandas udf?

**Comment by yma11** [comment]
@zhztheplayer Please help take a look again. Thanks.

File: gluten-data/src/main/java/org/apache/gluten/vectorized/ArrowWritableColumnVector.java [hunk 4]
```diff
@@ -345,6 +345,10 @@ public void retain() {
     refCnt.getAndIncrement();
   }
 
+  public AtomicLong getRefCntObj() {
+    return refCnt;
+  }
```
Comment (zhztheplayer) [inline gluten-data/src/main/java/org/apache/gluten/vectorized/ArrowWritableColumnVector.java:350]: It still looks risky... Let me add a fix on your PR

**Review by zhztheplayer — APPROVED** [review]

**Pull request merged**